### ISSUE TITLE
Web Worker Hangs if Imported File Contains Top-Level `Await`

### ISSUE THREAD

**Issue by asharonbaltazar** [GoogleChromeLabs/comlink#635]
Whatever your thoughts are on top-level await, I'm initializing a library using `await`. However, the web worker seems to hang once it encounters the `await`, preventing the rest of the web worker from executing. Any imported file with a top-level await causes this. 

I've made a [reproducible example](https://codesandbox.io/p/sandbox/silly-sun-hz4zsj). If the `count is {count}` button is clicked, the `getNumberPlusOne` function in `fakeTopLevelAwait` won't work. If the await is commented out, `getNumberPlusOne` executes successfully. 

I'd like to know if this a limitation in Web Workers or comlink. I'd really like to get this working. And kudos on such a fantastic lib!

**Comment by AhmetKaanGuney** [comment]
I had the exact same issue. I think that's because of how `import` is implemented for the worker.

**Comment by surma** [comment]
Right, so currently it’s the user’s responsibility to make sure the worker is ready before using Comlink. Because you create the worker in the click hander (side-note: generally I wouldn’t recommend that!), you call `getNumberPlusOne()` before the worker is ready. So the message that tells the worker that the function was called just gets lost.

You either need to send a ”ready!” message and wait for that first or — the more fragile, hackier and easier solution — is to move the worker creation to the top of the file. Unless the user clicks the button in under 2 seconds (because of your `setTimeout`) it works! See [here](https://codesandbox.io/p/sandbox/stoic-meadow-hty35s).

This is something I’d like to address in an update to Comlink where `wrap()` returns a Promise that you have to await where Comlink waits until the worker calls `expose()`.

**Comment by ansemjo** [comment]
> This is something I’d like to address in an update to Comlink where `wrap()` returns a Promise that you have to await where Comlink waits until the worker calls `expose()`.

I'd love that! I'm just beginning to use comlink-wrapped [Workers in a Vue component](https://vitejs.dev/guide/features.html#web-workers) and was wondering why the very simple `ready() { return true; }` function simply seemed to hang. Because the top-level [`<script setup>` block in single-file components](https://vuejs.org/api/sfc-script-setup.html) does not allow asynchronous functions, I've put the calls in the `onMounted` lifecycle hook. But the Promises only started resolving after I inserted a delay of several hundred milliseconds.

If the `Comlink.wrap()` function returns a Promise, I won't be able to use it in the `<script setup>` block but maybe I'm better off putting it all outside a component anyway. At least then I'll have a sure way of knowing the link is ready. 😊

**Comment by mccare** [comment]
Just had the same issue with vue and creating and setting up a worker in the setup section of my component. Here is the above solution in Javascript. 

```javascript
async function wrap(worker) {
     await new Promise((resolve, reject) => {
          const controller = new AbortController();
          worker.addEventListener('message', (message) => {
               if (message?.data?.ready) {
                    controller.abort();
                    resolve();
               }
          }, { signal: controller.signal })
     })
     return Comlink.wrap(worker);
}
```

In your worker:
```javascript
Comlink.expose(object);
postMessage({ready: true})
```

And to use it:
```javascript
import Worker from './worker?worker'

await wrap(new Worker());
```

**Comment by vinayakkulkarni** [comment]
Any repro Vue example?
